prope/prope.go

```diff
@@ -0,0 +1,275 @@
+// Package prope provides an implementation of a persistent rope data structure.
+// It is similar to the base rope data structure, but the changes
+// are saved separately without modifying the original data structure by
+// sharing data between multiple versions. The time complexity of operations
+// stay the same, but they are generally a bit slower:
+//
+// * Remove: O(lg n).
+//
+// * Insert: O(lg n).
+//
+// * Slice: O(lg n + m), where m is the size of the slice.
+//
+// * At: O(lg n).
+//
+// The main difference is in space complexity, as the persistent data structure
+// allows creating a copy with an insertion or removal in O(lg n) space,
+// instead of duplicating the entire rope for each change.
```

Comment:
I think this should specify that the space complexity gain is in comparison to copying and modifying the base rope (to get the same persistence guarantees). The persistent rope won't be more space efficient than just directly mutating the base rope (which is what insertion/removal for the base rope does), but of course this breaks the persistence property.